app/models/ems_refresh/save_inventory_physical_infra.rb

```diff
@@ -37,17 +36,23 @@ def save_ems_physical_infra_inventory(ems, hashes, target = nil)
     ems
   end
 
+  def save_physical_racks_inventory(ems, hashes, target = nil)
+    target = ems if target.nil?
+
+    deletes = target == ems ? :use_association : []
+
+    ems.physical_racks.reset
+    save_inventory_multi(ems.physical_racks, hashes, deletes, [:ems_ref], [:physical_servers])
```

Comment:
I feel like there's going to be a clash here where the physical_servers table is going to be updated via two different paths with two different sets of data...1) directly from the EMS, and 2) through this relationship, which will probably cause thrashing.

Reply:
@agrare How do we handle this with other tables?  I thought we only did the one save from the EMS, but I'm not sure.

Reply:
So yeah you're gong to have problems:
```
save_physical_servers([server1, server2])
save_physical_racks(rack)
  # rack has child server 3
  save_physical_servers([server3])
^^ is going to disconnect all servers _not_ in this rack
```

Reply:
@felipedf instead of having `:physical_servers` as a child of both the ems and the rack, just have all physical servers in the top level collection and reference the rack from the physical server.

Example:
```
:racks => [{:ems_ref => "1"}],
:physical_servers => [{:name => 'server1', :rack => {:ems_ref => "1"}}, {:name => "server1"}]
```

Save the racks first, then in save_physical_servers pull the new rack id like this:

`hash[:rack_id] = hash.delete(:rack).try(:id)`

This is assuming there is a rack_id on the physical_servers table

Reply:
I ran into this issue, what I tried at first was saving racks and then the servers associated to it, then I tried to save all servers related to the provider.
result: All servers previously added through the rack association were wiped out.

Then I tried to do the opposite, save all servers associated to the provider first, then the ones associated to the rack.
result: they were all there. but one of the behaviors we do want is when I try to access the association provider.physical_servers it will retrieve me all servers managed by that provider, the ones that are inside a rack, chassis, or alone. I tried few thing without success and the solution I found was something like server[ems_id] = ems_id while doing a parse.

@agrare I got to say that I don't know why this trashing happens while use associations in that way. Maybe you can give a hand here. About your suggestion, how would know which server is in which rack using that approach?

Reply:
You don't want to be saving the same type from multiple places, so instead of
>  save all servers associated to the provider first, then the ones associated to the rack.

You should just save all servers.

It belongs to a physical rack so you can put a physical_rack_id in a server record, but that doesn't mean that you should be saving those servers from the physical_rack association.

Reply:
That will work, but that rack_id should be set during the parse right?

Reply:
Sort-of, the actual id will be set at save time, the parser will link the server to the rack hash.

so if you have something like this in your parser:
```
def parse
  racks, rack_uids = parse_racks(inv[:racks])
  servers = parse_servers(inv[:servers], rack_uids)
end

def parse_racks(inv)
  results = []
  uids = {}

  inv.each do |rack|
    new_result = {
      // parse your rack here
    }

    results << new_result
    uids[new_result[:ems_ref]] = new_result
  end

  return results, uids
end

def parse_servers(inv, rack_uids)
  results = []

  inv.each do |server|
    rack = rack_uids[server[:rack_unique_identifier]]

    new_result = {
      :rack => rack,
    }
    results << new_result
  end

  results
end
```

This way when you save the racks collection first and update the IDs in the rack hashes you can then get the rack ID when you get to the servers.

Reply:
We should treat this similar to how we handle the host -> physical_server relationship.  The physical_infra provider only saves physical servers and if there is a host that matches the appropriate query; we create an association.  In the rack case, we should create all of the racks first.    When we parse the physical_server; if there is an appropriate rack instance we create the association at that time.   

Rack parsing should not create any physical_server objects. The rack parsing should only create the rack objects.     The physical_server parsing should create the association if a rack exists.